README
NAME
    Plack::Response::Data - Response Content Body Encoder for JSON, XML,
    YAML, etc.

SYNOPSIS
        use Plack::Response::Data;

        sub psgi_handler {
            my $env  = shift;
            my $req  = Plack::Request->new($env);
    
            my $data = { key => 'value', array => [1, 2, 3] };
            my $res  = Plack::Response::Data->new(200)->format('JSON')->data($data);
            return $res->finalize;
        }

DESCRIPTION
    This uses Plack::Response's response content body as a data container to
    respond a data encoded in JSON, XML, YAML format, etc. This would help
    you to build an Web API server application in a short code.

METHODS
    This is a subclass of "Plack::Response" class. This provides the
    following additional methods as well as other methods inherited since
    "Plack::Response".

  format
    This specifies a data container format. Any formats found at
    "Plack::Response::Data::*" modules are supported.

        $res->format('JSON');       # uses Plack::Response::Data::JSON
        $res->format('XML');        # uses Plack::Response::Data::XML
        $res->format('YAML');       # uses Plack::Response::Data::YAML

    Without arguments given, this returns the current format.

        $format = $res->format;     # returns 'JSON' etc.

  data
    This writes a data object into response content body.

        $data = {key => 'val'};     # a hashref
        $res->data($data);
        $body = $res->body;         # returns a JSON encoded string: '{"key":"val"}'

    Without arguments given, this parses the current resopnse content body
    as a data object vice versa.

        $body = '{"key":"val"}';    # a JSON encoded string
        $res->body($body);
        $data = $res->data;         # returns a hashref decoded: {key => 'val'}

AUTHOR
    Yusuke Kawasaki

SEE ALSO
    Plack::Response

    Plack::Response::Data::JSON

    Plack::Response::Data::XML

    Plack::Response::Data::YAML



Changes
Sun Apr 22 08:31:56 PDT 2012
    - first test


Makefile.PL
use ExtUtils::MakeMaker;

my $opt = {
    NAME                =>  'Plack::Response::Data',
    VERSION_FROM        =>  'lib/Plack/Response/Data.pm',
    ABSTRACT_FROM       =>  'lib/Plack/Response/Data.pm',
    CONFIGURE_REQUIRES  =>  {
        'ExtUtils::MakeMaker'   => '6.55',
    },
    BUILD_REQUIRES      =>  {
        'Test::More'            => '0.71',
        'Test::Pod'             => '1.41',
    },
    PREREQ_PM           =>  {
        'Carp'                  => '0',
        'JSON'                  => '2.00',
        'HTTP::Status'          => '6.00',
        'Plack'                 => '0.9948',
        'XML::TreePP'           => '0.41',
        'YAML::Any'             => '0.67',
    },
    AUTHOR              =>  'Yusuke Kawasaki',
    LICENSE             =>  'perl',
};

WriteMakefile( %$opt );


lib/Plack/Response/Data/XML.pm
package Plack::Response::Data::XML;
use strict;
use warnings;
use base 'Plack::Response::Data';

our $VERSION = '0.01';
$VERSION = eval $VERSION;

use Carp ();
use XML::TreePP;

sub mime_type {
    'application/xml';
}

sub handler {
    my $handler = XML::TreePP->new;
    $handler;
}

sub encode {
    my $self = shift;
    my $data = shift;
    my $handler = $self->handler or Carp::croak "XML handler was not available";
    my $body = $handler->write($data);
    utf8::encode($body);
    $body;
}

sub decode {
    my $self = shift;
    my $body = shift;
    my $handler = $self->handler or Carp::croak "XML handler was not available";
    utf8::decode($body);
    my $data = $handler->parse($body);
    $data;
}


1;
__END__

=head1 NAME

Plack::Response::Data::XML - Response Content Body Encoder/Decoder for XML

=head1 SYNOPSIS

In your Plack app:

    use Plack::Response::Data::XML;

    sub psgi_handler {
        my $env  = shift;
        my $req  = Plack::Request->new($env);
    
        my $data = { foo => 'bar', qux => [1, 2, 3] };
        my $res  = Plack::Response::Data::XML->new(200)->data($data);
        return $res->finalize;
    }

Response:

    HTTP/1.0 200 OK
    Content-Length: 107
    Content-Type: application/xml
    
    <?xml version="1.0" encoding="UTF-8" ?>
    <xml>
    <foo>bar</foo>
    <qux>1</qux>
    <qux>2</qux>
    <qux>3</qux>
    </xml>

=head1 DESCRIPTION

This uses L<Plack::Response>'s response content body as a data container
to respond a data encoded in XML format.
This would help you to build an Web API server application in a short code.
L<XML::TreePP> module is required.

=head1 METHODS

This is a subclass of C<Plack::Response::Data> class.

=head2 data

This writes a data object into the response content body as XML.

    $data = {foo => 'bar'};     # a hashref
    $res->data($data);
    $body = $res->body;         # returns an XML encoded string: '{"foo":"bar"}'

Without arguments given, this parses the current resopnse content body
as a data object vice versa.

    $body = '{"foo":"bar"}';    # an XML encoded string
    $res->body($body);
    $data = $res->data;         # returns a hashref decoded: {foo => 'bar'}

=head2 encode, decode

This also provides the bidirectional encoder and decoder methods as below.

    $text = Plack::Response::Data::XML->encode($data); # object to string
    $data = Plack::Response::Data::XML->decode($text); # string to object

=head1 SAMPLE ECHO APP

The following PSGI app does just echo an XML request to an XML response.
You would need error handlings and security considerations for production
use however.

    use Plack::Request;
    use Plack::Response::Data::XML;
    
    sub {
        my $env  = shift;
        my $req  = Plack::Request->new($env);
        my $body = $req->content;
        my $data = Plack::Response::Data::XML->decode($body);
        # do something
        my $res  = Plack::Response::Data::XML->new(200);
        $res->data($data);
        $res->finalize;
    }

=head1 AUTHOR

Yusuke Kawasaki

=head1 SEE ALSO

L<Plack::Response>

L<XML::TreePP>

=cut


lib/Plack/Response/Data/JSON.pm
package Plack::Response::Data::JSON;
use strict;
use warnings;
use base 'Plack::Response::Data';

our $VERSION = '0.01';
$VERSION = eval $VERSION;

use Carp ();
use JSON ();

sub mime_type {
    'application/json';
}

sub handler {
    my $handler = JSON->new;
    $handler->utf8(1);
    $handler;
}

sub encode {
    my $self = shift;
    my $data = shift;
    my $handler = $self->handler or Carp::croak "JSON handler was not available";
    my $body;
    eval {
        $body = $handler->encode($data);
    };
    $body;
}

sub decode {
    my $self = shift;
    my $body = shift;
    my $handler = $self->handler or Carp::croak "JSON handler was not available";
    my $data;
    eval {
        $data = $handler->decode($body);
    };
    $data;
}

1;
__END__

=head1 NAME

Plack::Response::Data::JSON - Response Content Body Encoder/Decoder for JSON

=head1 SYNOPSIS

In your Plack app:

    use Plack::Response::Data::JSON;

    sub psgi_handler {
        my $env  = shift;
        my $req  = Plack::Request->new($env);
    
        my $data = { foo => 'bar', qux => [1, 2, 3] };
        my $res  = Plack::Response::Data::JSON->new(200)->data($data);
        return $res->finalize;
    }

Response:

    HTTP/1.0 200 OK
    Content-Length: 13
    Content-Type: application/json
    
    {"foo":"bar","qux":[1,2,3]}

=head1 DESCRIPTION

This uses L<Plack::Response>'s response content body as a data container
to respond a data encoded in JSON.
This would help you to build an Web API server application in a short code.
L<JSON> module is required.

=head1 METHODS

This is a subclass of C<Plack::Response::Data> class.

=head2 data

This writes a data object into the response content body as JSON.

    $data = {foo => 'bar'};     # a hashref
    $res->data($data);
    $body = $res->body;         # returns a JSON encoded string: '{"foo":"bar"}'

Without arguments given, this parses the current resopnse content body
as a data object vice versa.

    $body = '{"foo":"bar"}';    # a JSON encoded string
    $res->body($body);
    $data = $res->data;         # returns a hashref decoded: {foo => 'bar'}

=head2 encode, decode

This also provides the bidirectional encoder and decoder methods as below.

    $text = Plack::Response::Data::JSON->encode($data); # object to string
    $data = Plack::Response::Data::JSON->decode($text); # string to object

=head1 SAMPLE ECHO APP

The following PSGI app does just echo a JSON request to a JSON response.
You would need error handlings and security considerations for production
use however.

    use Plack::Request;
    use Plack::Response::Data::JSON;
    
    sub {
        my $env  = shift;
        my $req  = Plack::Request->new($env);
        my $body = $req->content;
        my $data = Plack::Response::Data::JSON->decode($body);
        # do something
        my $res  = Plack::Response::Data::JSON->new(200);
        $res->data($data);
        $res->finalize;
    }

=head1 AUTHOR

Yusuke Kawasaki

=head1 SEE ALSO

L<Plack::Response>

L<JSON>

=cut


lib/Plack/Response/Data/YAML.pm
package Plack::Response::Data::YAML;
use strict;
use warnings;
use base 'Plack::Response::Data';

our $VERSION = '0.01';
$VERSION = eval $VERSION;

use Carp ();
use YAML::Any;

sub mime_type {
    'text/yaml';
}

sub encode {
    my $self = shift;
    my $data = shift;
    my $body = Dump($data);
    $body;
}

sub decode {
    my $self = shift;
    my $body = shift;
    my $data = Load($body);
    $data;
}

1;
__END__

=head1 NAME

Plack::Response::Data::YAML - Response Content Body Encoder/Decoder for YAML

=head1 SYNOPSIS

In your Plack app:

    use Plack::Response::Data::YAML;

    sub psgi_handler {
        my $env  = shift;
        my $req  = Plack::Request->new($env);

        my $data = { foo => 'bar', qux => [1, 2, 3] };
        my $res  = Plack::Response::Data::YAML->new(200)->data($data);
        return $res->finalize;
    }

Response:

    HTTP/1.0 200 OK
    Content-Length: 38
    Content-Type: text/yaml
    
    --- 
    foo: bar
    qux: 
      - 1
      - 2
      - 3

=head1 DESCRIPTION

This uses L<Plack::Response>'s response content body as a data container
to respond a data encoded in YAML.
L<YAML::Any> module is required.

=head1 METHODS

This is a subclass of C<Plack::Response::Data> class.

=head2 data

This writes a data object into the response content body as YAML.

    $data = {foo => 'bar'};     # a hashref
    $res->data($data);
    $body = $res->body;         # returns a YAML encoded string

Without arguments given, this parses the current resopnse content body
as a data object vice versa.

    $body = '{"foo":"bar"}';    # a YAML encoded string
    $res->body($body);
    $data = $res->data;         # returns a hashref decoded

=head2 encode, decode

This also provides the bidirectional encoder and decoder methods as below.

    $text = Plack::Response::Data::YAML->encode($data); # object to string
    $data = Plack::Response::Data::YAML->decode($text); # string to object

=head1 AUTHOR

Yusuke Kawasaki

=head1 SEE ALSO

L<Plack::Response>

L<YAML::Any>

=cut


lib/Plack/Response/Data/Status.pm
package Plack::Response::Data::Status;
use strict;
use warnings;
use base 'Plack::Response::Data';

use HTTP::Status ();

sub mime_type {
    'text/plain';
}

sub encode {
    my $self = shift;
    my $code = shift or return;
    my $mess = HTTP::Status::status_message($code);
    $mess =~ s/\s*$/\015\012/s;
    $mess;
}

sub code {
    my $self = shift;
    return $self->SUPER::code unless @_;
    $self->SUPER::data(@_);
    $self->SUPER::code(@_);
}

sub status {
    my $self = shift;
    return $self->SUPER::status unless @_;
    $self->SUPER::data(@_);
    $self->SUPER::status(@_);
}

sub data {
    my $self = shift;
    return $self->SUPER::status unless @_;
    $self->SUPER::status(@_);
    $self->SUPER::data(@_);
}

sub body {
    my $self = shift;

    # setter
    return $self->SUPER::body(@_) if @_;

    # getter
    my $body = $self->SUPER::body;
    return $body if defined $body;

    # default
    my $code = $self->status;
    $self->data($code) if $code;

    $self->SUPER::body;
}

sub finalize {
    my $self = shift;
    $self->body; # touch
    $self->SUPER::finalize(@_);
}

1;
__END__

=head1 NAME

Plack::Response::Data::Status - Response Content Body of Status Message

=head1 SYNOPSIS

In your Plack app:

    use Plack::Response::Data::Status;

    sub psgi_handler {
        my $env  = shift;
        my $req  = Plack::Request->new($env);
     
        my $res  = Plack::Response::Data::Status->new(404);
        return $res->finalize;
    }

Response:

    HTTP/1.0 404 Not Found
    Content-Length: 11
    Content-Type: text/plain
    
    Not Found

=head1 DESCRIPTION

This writes the HTTP response status message at L<Plack::Response>'s response content body.
This would help you to respond an error code with its reason phrase
which is a short textual description defined in RFC 2616 and RFC 2518.
Response headers of C<Content-Type> and C<Content-Length> are also given.
L<HTTP::Status> module is required.

=head1 METHODS

This is a subclass of C<Plack::Response::Data> class.

=head2 code, data, status

The three methods effect the content body in the same manner.

    $res->code(404);
    $res->data(404);
    $res->status(404);

=head1 AUTHOR

Yusuke Kawasaki

=head1 SEE ALSO

L<Plack::Response>

L<Status>

=cut


lib/Plack/Response/Data.pm
package Plack::Response::Data;
use strict;
use warnings;
use base 'Plack::Response';

our $VERSION = '0.01';
$VERSION = eval $VERSION;

use Carp ();
use Plack::Util;

sub mime_type {
    Carp::croak "Not implemented";
}

sub encode {
    Carp::croak "Not implemented";
}

sub decode {
    Carp::croak "Not implemented";
}

sub data {
    my $self = shift;

    if (@_) {
        my $data = shift;
        my $body = $self->encode($data);
        $self->body($body);
        my $type = $self->mime_type;
        $self->content_type($type) if $type;
        $self->content_length(length($body));
        return $self; # chained
    }

    my $body = $self->_body;
    return unless defined $body;

    if (ref $body ne 'ARRAY') {
        $body = join("" => @$body);
    } elsif (ref $body) {
        my @buf;
        Plack::Util::foreach($body, sub { push(@buf, $_[0]) });
        $self->body(\@buf);
        $body = join("" => @buf);
    }
    $self->decode($body);
}

sub format {
    my $self = shift;
    $self = $self->new unless ref $self;

    my $current = ref $self;
    my $base    = __PACKAGE__;
    my $format;

    if (@_) {
        $format = shift or Carp::croak "Invalid container format";
        my $class  = $base.'::'.$format;
        if ($class ne $current) {
            my $file = $class;
            $file =~ s!::!/!g;
            $file .= '.pm';
            $class = Plack::Util::load_class($format, $base) unless $INC{$file};
            bless $self, $class;
        }
        $self->body(undef); # empty
        return $self; # chained
    }

    $format = ($current =~ m/^\Q$base\E::(.*)$/)[0];
    $format;
}

sub _get_version {
    my $class = shift;
    my $name = $class."::VERSION";
    no strict 'refs';
    $$name;
}

1;
__END__

=head1 NAME

Plack::Response::Data - Response Content Body Encoder for JSON, XML, YAML, etc.

=head1 SYNOPSIS

    use Plack::Response::Data;

    sub psgi_handler {
        my $env  = shift;
        my $req  = Plack::Request->new($env);
    
        my $data = { key => 'value', array => [1, 2, 3] };
        my $res  = Plack::Response::Data->new(200)->format('JSON')->data($data);
        return $res->finalize;
    }

=head1 DESCRIPTION

This uses L<Plack::Response>'s response content body as a data container
to respond a data encoded in JSON, XML, YAML format, etc.
This would help you to build an Web API server application in a short code.

=head1 METHODS

This is a subclass of C<Plack::Response> class.
This provides the following additional methods
as well as other methods inherited since C<Plack::Response>.

=head2 format

This specifies a data container format.
Any formats found at C<Plack::Response::Data::*> modules are supported.

    $res->format('JSON');       # uses Plack::Response::Data::JSON
    $res->format('XML');        # uses Plack::Response::Data::XML
    $res->format('YAML');       # uses Plack::Response::Data::YAML

Without arguments given, this returns the current format.

    $format = $res->format;     # returns 'JSON' etc.

=head2 data

This writes a data object into response content body.

    $data = {key => 'val'};     # a hashref
    $res->data($data);
    $body = $res->body;         # returns a JSON encoded string: '{"key":"val"}'

Without arguments given, this parses the current resopnse content body
as a data object vice versa.

    $body = '{"key":"val"}';    # a JSON encoded string
    $res->body($body);
    $data = $res->data;         # returns a hashref decoded: {key => 'val'}

=head1 AUTHOR

Yusuke Kawasaki

=head1 SEE ALSO

L<Plack::Response>

L<Plack::Response::Data::JSON>

L<Plack::Response::Data::XML>

L<Plack::Response::Data::YAML>

=cut
